ams/server/src/main/java/com/netease/arctic/server/table/executor/DataExpiringExecutor.java

```diff
@@ -0,0 +1,468 @@
+        CloseableIterable<FileEntry> mayExpiredFiles = CloseableIterable.withNoopClose(
+            Lists.newArrayList(CloseableIterable.filter(entries,
+                e -> mayExpired(e, expirationConfig, partitionFreshness, expireTimestamp))));
+        CloseableIterable.filter(mayExpiredFiles, e -> willNotRetain(e, expirationConfig, partitionFreshness))
+            .forEach(e -> {
+              if (e.isChange) {
+                changeExpiredFiles.addFile(e);
+              } else {
+                baseExpiredFiles.addFile(e);
+              }
+            });
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      expireFiles(changeTable, changeSnapshot.snapshotId(), changeExpiredFiles, expireTimestamp);
+      expireFiles(baseTable, baseSnapshot.snapshotId(), baseExpiredFiles, expireTimestamp);
+    } else {
+      UnkeyedTable unkeyedTable = table.asUnkeyedTable();
+      Snapshot snapshot = unkeyedTable.currentSnapshot();
+      ExpireFiles expiredFiles = new ExpireFiles();
+      try (CloseableIterable<IcebergFileEntry> entries
+          = fileEntriesScan(unkeyedTable, partitionFilter, snapshot).entries()) {
+        CloseableIterable<IcebergFileEntry> mayExpiredFiles = CloseableIterable.withNoopClose(
+            Lists.newArrayList(CloseableIterable.filter(entries,
+                e -> mayExpired(e, expirationConfig, partitionFreshness, expireTimestamp))));
+        CloseableIterable.filter(
+                mayExpiredFiles,
+                e -> willNotRetain(e, expirationConfig, partitionFreshness))
+            .forEach(expiredFiles::addFile);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      expireFiles(unkeyedTable, snapshot.snapshotId(), expiredFiles, expireTimestamp);
+    }
+  }
+
+  private static Expression getPartitionExpression(DataExpirationConfig expirationConfig, long expireTimestamp) {
+    if (expirationConfig.expirationLevel.equals(ExpireLevel.PARTITION)) {
+      return Expressions.alwaysTrue();
+    }
+
+    Type.TypeID typeID = expirationConfig.expirationField.type().typeId();
+    if (typeID == Type.TypeID.TIMESTAMP) {
+      return Expressions.lessThanOrEqual(expirationConfig.expirationField.name(), expireTimestamp * 1000);
+    }
+    return Expressions.alwaysTrue();
+  }
+
+  private static void expireFiles(UnkeyedTable table, long snapshotId, ExpireFiles expiredFiles, long expireTimestamp) {
+    List<DataFile> dataFiles = expiredFiles.dataFiles;
+    List<DeleteFile> deleteFiles = expiredFiles.deleteFiles;
+    if (dataFiles.isEmpty() && deleteFiles.isEmpty()) {
+      return;
+    }
+    // expire data files
+    DeleteFiles delete = table.newDelete();
+    dataFiles.forEach(delete::deleteFile);
+    delete.set(SnapshotSummary.SNAPSHOT_PRODUCER, "Amoro data expiration");
+    delete.commit();
+    // expire delete files
+    if (!deleteFiles.isEmpty()) {
```

Comment:
Is it better to apply both operations, Delete datafile and DeleteFile rewrite, in one transaction?

Reply:
Do you mean replace thoses operation with RewriteFiles interface?